### PR TITLE
[api] Add virtual QgsRasterRenderer::setInputBand method

### PR DESCRIPTION
**Pull request by nyalldawson**
Attempts to set the input band for the renderer.
Returns TRUE if the band was successfully set, or FALSE if the band could not be set.

This was implemented in various raster renderer subclasses, but it was necessary to down cast and then call the individual methods (which don't have consistent names :scream:).

Instead, add a top level virtual method so that it's easy to change the input band for the renderers.

### CODE REVIEW

File: src/core/raster/qgssinglebandpseudocolorrenderer.cpp [hunk 1]
```diff
@@ -37,17 +37,23 @@ QgsSingleBandPseudoColorRenderer::QgsSingleBandPseudoColorRenderer( QgsRasterInt
...
   if ( !mInput )
   {
-    mBand = bandNo;
-    return;
+    mBand = band;
+    return true;
   }
-
-  if ( bandNo <= mInput->bandCount() || bandNo > 0 )
+  else if ( band <= mInput->bandCount() || band > 0 )
```
Comment (rouault) [inline src/core/raster/qgssinglebandpseudocolorrenderer.cpp:51]: suspicious condition... likely
```suggestion
  else if ( band <= mInput->bandCount() && band > 0 )
```

File: src/core/raster/qgspalettedrasterrenderer.cpp [hunk 2]
```diff
@@ -175,6 +175,12 @@ void QgsPalettedRasterRenderer::setLabel( double idx, const QString &label )
   }
 }
 
+bool QgsPalettedRasterRenderer::setInputBand( int band )
+{
+  mBand = band;
```
Comment (rouault) [inline src/core/raster/qgspalettedrasterrenderer.cpp:180]: no checking on band values needed?

File: src/core/raster/qgsrastercontourrenderer.cpp [hunk 3]
```diff
@@ -232,6 +232,12 @@ QList<QgsLayerTreeModelLegendNode *> QgsRasterContourRenderer::createLegendNodes
   return nodes;
 }
 
+bool QgsRasterContourRenderer::setInputBand( int band )
+{
+  mInputBand = band;
```
Comment (rouault) [inline src/core/raster/qgsrastercontourrenderer.cpp:237]: no checking on band values needed?

File: src/core/raster/qgssinglebandcolordatarenderer.cpp [hunk 4]
```diff
@@ -139,3 +140,9 @@ bool QgsSingleBandColorDataRenderer::setInput( QgsRasterInterface *input )
   }
   return false;
 }
+
+bool QgsSingleBandColorDataRenderer::setInputBand( int band )
+{
+  mBand = band;
```
Comment (rouault) [inline src/core/raster/qgssinglebandcolordatarenderer.cpp:146]: no checking on band values needed?

File: src/core/raster/qgssinglebandgrayrenderer.cpp [hunk 5]
```diff
@@ -193,6 +193,17 @@ QgsRasterBlock *QgsSingleBandGrayRenderer::block( int bandNo, const QgsRectangle
...
 }
 
+void QgsSingleBandGrayRenderer::setGrayBand( int band )
+{
+  setInputBand( band );
+}
+
+bool QgsSingleBandGrayRenderer::setInputBand( int band )
+{
+  mGrayBand = band;
```
Comment (rouault) [inline src/core/raster/qgssinglebandgrayrenderer.cpp:203]: no checking on band values needed?

**Review by rouault — COMMENTED** [review]
Perhaps add a minimal test of the new API ?

**Comment by nyalldawson** [comment]
@rouault ok, all addressed and a bunch of new tests added!

**Review by wonder-sk — APPROVED** [review]

**Pull request merged**